File: backend/src/routes/stocks.js
import { Hono } from "hono";
import { getHistory, getQuote, searchStocks, UnknownTickerError } from "../market.js";
import { ApiError } from "../util.js";

const VALID_PERIODS = new Set(["5d", "1mo", "3mo", "6mo", "1y", "2y", "5y", "max"]);

const router = new Hono();

router.get("/search", async (c) => {
  const query = (c.req.query("q") || "").trim();
  if (!query || query.length > 50) throw new ApiError(422, "q must be 1-50 characters");
  return c.json(await searchStocks(query));
});

router.get("/:ticker/quote", async (c) => {
  const ticker = c.req.param("ticker").toUpperCase().trim();
  try {
    return c.json({ ticker, price: await getQuote(ticker) });
  } catch (error) {
    if (error instanceof UnknownTickerError) throw new ApiError(404, error.message);
    throw error;
  }
});

router.get("/:ticker/history", async (c) => {
  const period = c.req.query("period") || "1mo";
  if (!VALID_PERIODS.has(period)) {
    throw new ApiError(400, `period must be one of ${[...VALID_PERIODS].sort().join(", ")}`);
  }
  try {
    return c.json(await getHistory(c.req.param("ticker"), period));
  } catch (error) {
    if (error instanceof UnknownTickerError) throw new ApiError(404, error.message);
    throw error;
  }
});

export default router;

File: backend/app/routers/stocks.py
from fastapi import APIRouter, HTTPException, Query

from app.schemas import HistoryPoint, QuoteOut, StockSearchResult
from app.services import market
from app.services.market import UnknownTickerError

router = APIRouter(prefix="/stocks", tags=["stocks"])

VALID_PERIODS = {"5d", "1mo", "3mo", "6mo", "1y", "2y", "5y", "max"}


@router.get("/search", response_model=list[StockSearchResult])
def search_stocks(q: str = Query(min_length=1, max_length=50)):
    return market.search(q)


@router.get("/{ticker}/quote", response_model=QuoteOut)
def get_quote(ticker: str):
    try:
        return QuoteOut(ticker=ticker.upper(), price=market.get_quote(ticker))
    except UnknownTickerError as exc:
        raise HTTPException(404, str(exc))


@router.get("/{ticker}/history", response_model=list[HistoryPoint])
def get_history(ticker: str, period: str = "1mo"):
    if period not in VALID_PERIODS:
        raise HTTPException(400, f"period must be one of {sorted(VALID_PERIODS)}")
    try:
        return market.get_history(ticker, period)
    except UnknownTickerError as exc:
        raise HTTPException(404, str(exc))
